### PR TITLE
Added Azure.VMSS.PublicIPAttached

### LINKED ISSUE #3014: [RULE] Avoid attaching public IPs directly to virtual machine scale set instances
### Existing rule

_No response_

### Suggested rule

Attaching a public IP address to a virtual machine network interface (NIC) exposes it directly to the Internet.
This exposure can make the virtual machine scale set instance vulnerable to unauthorized inbound access and security compromise.
Minimize the number of Internet ingress/ egress points to enhance security and reduces potential attack surfaces.

### Pillar

Security

### Additional context

- [SE:06 Network controls](https://learn.microsoft.com/azure/well-architected/security/networking)
- [Plan for inbound and outbound internet connectivity](https://learn.microsoft.com/azure/cloud-adoption-framework/ready/azure-best-practices/plan-for-inbound-and-outbound-internet-connectivity)
- [Networking for scale sets](https://learn.microsoft.com/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-networking)
- [Public IPv4 per virtual machine](https://learn.microsoft.com/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-networking#public-ipv4-per-virtual-machine)
- [Azure Bastion](https://learn.microsoft.com/azure/bastion/bastion-overview)
- [Azure deployment reference - VMSS](https://learn.microsoft.com/azure/templates/microsoft.compute/virtualmachinescalesets)
- [Azure deployment reference - VMSS instance](https://learn.microsoft.com/azure/templates/microsoft.compute/virtualmachinescalesets/virtualmachines)
- [Azure deployment reference - NIC](https://learn.microsoft.com/azure/templates/microsoft.network/networkinterfaces)

### PR DESCRIPTION
**Pull request by BenjaminEngeset**
## PR Summary

Fixes #3014

Added Azure.VMSS.PublicIPAttached.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: tests/PSRule.Rules.Azure.Tests/Azure.VMSS.Tests.ps1 [hunk 1]
```diff
@@ -183,8 +183,27 @@ Describe 'Azure.VMSS' -Tag 'VMSS' {
...
+
+            $ruleResult[0].Reason | Should -BeExactly "The virtual machine scale set instances should not have public IP addresses directly attached to their network interfaces.";
+            $ruleResult[1].Reason | Should -BeExactly "The virtual machine scale set instances should not have public IP addresses directly attached to their network interfaces.";
+            $ruleResult[2].Reason | Should -BeExactly "The virtual machine scale set instances should not have public IP addresses directly attached to their network interfaces.";
+            $ruleResult[3].Reason | Should -BeExactly "The virtual machine scale set instances should not have public IP addresses directly attached to their network interfaces.";
+
+            # Pass
+            $ruleResult = @($filteredResult | Where-Object { $_.Outcome -eq 'Pass' });
+            $ruleResult.Length | Should -Be 5;
+            $ruleResult.TargetName | Should -Be 'vmss-003', 'vmss-004', 'vmss-005', 'vmss-006/instance-B', 'nic-B';
```
Comment (BernieWhite) [inline tests/PSRule.Rules.Azure.Tests/Azure.VMSS.Tests.ps1:206]: As mentioned in rule.


```suggestion
            $ruleResult.Length | Should -Be 3;
            $ruleResult.TargetName | Should -BeIn 'vmss-003', 'vmss-006/instance-A', 'vmss-006/instance-B';
        }

        It 'Azure.VMSS.PublicIPAttached' {
            $filteredResult = $result | Where-Object { $_.RuleName -eq 'Azure.VMSS.PublicIPAttached' };

            # Fail
            $ruleResult = @($filteredResult | Where-Object { $_.Outcome -eq 'Fail' });
            $ruleResult.Length | Should -Be 3;
            $ruleResult.TargetName | Should -Be 'vmss-001', 'vmss-002', 'vmss-006/instance-A';

            $ruleResult[0].Reason | Should -BeExactly "The virtual machine scale set instances should not have public IP addresses directly attached to their network interfaces.";
            $ruleResult[1].Reason | Should -BeExactly "The virtual machine scale set instances should not have public IP addresses directly attached to their network interfaces.";
            $ruleResult[2].Reason | Should -BeExactly "The virtual machine scale set instances should not have public IP addresses directly attached to their network interfaces.";

            # Pass
            $ruleResult = @($filteredResult | Where-Object { $_.Outcome -eq 'Pass' });
            $ruleResult.Length | Should -Be 4;
            $ruleResult.TargetName | Should -Be 'vmss-003', 'vmss-004', 'vmss-005', 'vmss-006/instance-B';
```

**Review by BernieWhite — CHANGES_REQUESTED** [review]
Thanks @BenjaminEngeset. Looks great.

I have a few suggestions below.

**Comment by BenjaminEngeset** [comment]
Requested changes have been implemented, @BernieWhite. Please check out my comment regarding your first suggestion.

I am happy with the rule now.

**Review by BernieWhite — APPROVED** [review]
Great work @BenjaminEngeset. All good to merge.

**Comment by BenjaminEngeset** [comment]
Thanks @BernieWhite. Based on this screenshot, could you please elaborate on this? 

image

**Pull request merged**